testsuite/tests/effects/dynamic.ml

```diff
@@ -0,0 +1,254 @@
+end
+
+type _ Effect.t += E : unit -> unit Effect.t
+
+(* `trigger ()` is `(wait, go)`, where `wait()` will wait until `go()` is called.
+    and resets the trigger (so go() can be meaningfully called more than once). *)
+
+let trigger () =
+  let t = Atomic.make false in
+  ((fun () -> (while not (Atomic.get t) do Thread.yield () done; Atomic.set t false)),
+   (fun () -> Atomic.set t true))
+
+(* `on_my_mark f` makes a thread and returns a function `go`. When `go()` is called, the
+   thread will call `f` and return; `go()` joins the thread before returning. *)
+
+let on_my_mark f =
+  let wait, go = trigger () in
+  let t = Thread.create (fun () -> (wait(); f())) () in
+  fun () -> (go (); Thread.join t)
+
+(* `sync_thread f` make a thread and returns a pair of functions `(go, stop)`.
+   Whenever `go()` is called, the other thread runs `f`, synchronously. When
+   `stop()` is called, the other thread stops and is joined. *)
+
+let sync_thread f =
+  let wait1, go1 = trigger () in
+  let wait2, go2 = trigger () in
+  let stop = Atomic.make false in
+  let t = Thread.create (fun () -> (while not (Atomic.get stop) do
+                                      wait1 ();
+                                      f ();
+                                      go2()
+                                    done)) () in
+  ((fun () -> (go1(); wait2())),
+   (fun () -> (Atomic.set stop true; go1(); Thread.join t)))
+
+(* Actual Dynamic.t tests from here on *)
+
+(* `get` should return the `original` value:
+   - on the same thread;
+   - on another thread;
+   - on a thread created before the dynamic itself.
+*)
+
+let _ =
+  let r = ref (Dynamic.make 57) in
+  let go = on_my_mark (fun () -> (Printf.printf "get on earlier thread [expect 1]: %d\n" (Dynamic.get (!r)))) in
+  let d = Dynamic.make 1 in
+  (Printf.printf "get [expect 1]: %d\n" (Dynamic.get d);
+   Thread.join (Thread.create (fun () -> Printf.printf "get on other thread [expect 1]: %d\n" (Dynamic.get d)) ());
+   r := d;
+   go ())
+
+(* `set_root` should change the value seen:
+   - on this thread;
+   - on a child thread created after `set_root`;
+   - but not on a child thread created previously.
+*)
+
+let _ =
+  let r = ref (Dynamic.make 57) in
+  let go = on_my_mark (fun () -> (Printf.printf "get on pre-existing thread after set_root [expect 3]: %d\n" (Dynamic.get (!r)))) in
+  let d = Dynamic.make 3 in
+  (Dynamic.set_root d 2;
+   Printf.printf "get after set_root [expect 2]: %d\n" (Dynamic.get d);
+   Thread.join (Thread.create (fun () -> Printf.printf "get on child thread created after set_root [expect 2]: %d\n" (Dynamic.get d)) ());
+   r := d;
+   go ();
+   Thread.join (Thread.create (fun () -> Dynamic.set_root d 4) ());
+   Printf.printf "get after set_root on child thread [expect 2]: %d\n" (Dynamic.get d))
+
+(* `with_temporarily` should change the value seen:
+   - within its dynamic extent;
+   - but not on some other thread while it is running;
+   - or after it returns;
```

Comment:
What about a thread created inside of `with_temporarily` (presumably no - is that what we want?)

Reply:
Do people want nasally-inserted fire? I genuinely don't know. This could be done by copying fiber binding values into the per-thread root bindings, at thread creation time. Leaving this as-is until somebody clarifies the requirement.